### ISSUE TITLE
Dynamic On-chain Commitment and Nullifier Arrays

### ISSUE THREAD

**Issue by IlyasRidhuan** [EYBlockchain/nightfall_3#986]
We are currently restricted to circuits that use up to 4 nullifiers and output 3 commitments. This unnecessary rigidity can be eliminated by making the arrays in the `Transaction Struct` dynamic in size ( note this does not change or update the behaviour of existing circuits).

### Security
Some investigation is required on whether there is an unsafe upper bound to the size of these dynamic arrays - in terms of prohibitive cost of challenge as transactions need to be passed in as part of challenges.

### Pricing
This change might require a change in the way the price of transactions are looked at - from a fixed cost to a `cost per byte`.

**Comment by druiz0992** [comment]
So maybe in addition to number of tx per block we may limit number of commitments and nullifiers?

**Comment by IlyasRidhuan** [comment]
Yep, **We may need to change the concept of `MAX_TX_PER_BLOCK` instead to `MAX_BYTES_PER_BLOCK`** . 

However, the scaling is way different - so this might not even be a problem realistically. Every new `tx per block` adds about 24 words - but each new commitment is only 1 word. 

Our limit in transactions is roughly 64txs per block (once the optimisation Pas are merged). This would be the same as having a single transaction with 750 commitments and 750 nullifiers - I doubt there would be a circuit that would be able to handle that.
